### ISSUE TITLE
lxc-usernsexec fails with 'Failed to find subuid or subgid allocation' with proper mappings

### ISSUE THREAD

**Issue by amarao** [lxc/lxc#3121]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of
   * `lxc-start --version`: 3.0.3
   * `lxc-checkconfig`
```
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-4.15.0-58-generic
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 
/sys/fs/cgroup/systemd
/sys/fs/cgroup/blkio
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/hugetlb
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/devices
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/memory
/sys/fs/cgroup/pids
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/rdma
/sys/fs/cgroup/freezer

Cgroup v2 mount points: 
/sys/fs/cgroup/unified

Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, not loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, not loaded
CONFIG_NF_NAT_IPV4: enabled, loaded
CONFIG_NF_NAT_IPV6: enabled, not loaded
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, not loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 
```
   * `uname -a`: `Linux bionic-amarao-2019-08-20 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
   * `cat /proc/self/cgroup`
```
12:freezer:/
11:rdma:/
10:perf_event:/
9:pids:/user.slice/user-1000.slice/session-1.scope
8:memory:/user.slice
7:cpuset:/
6:devices:/user.slice
5:cpu,cpuacct:/user.slice
4:hugetlb:/
3:net_cls,net_prio:/
2:blkio:/user.slice
1:name=systemd:/user.slice/user-1000.slice/session-1.scope
0::/user.slice/user-1000.slice/session-1.scope
```
   * `cat /proc/1/mounts`
```
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=482368k,nr_inodes=120592,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=100908k,mode=755 0 0
/dev/vda1 / ext4 rw,relatime,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=24,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=12034 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime,pagesize=2M 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
configfs /sys/kernel/config configfs rw,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=100904k,mode=700,uid=1000,gid=1001 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,relatime 0 0
```
# Issue description
`lxc-usernsexec` fails to find mapping even mapping is the proper format.
# Steps to reproduce

1.
```
cat /etc/subuid
lxd:165536:65536
root:165536:65536
cloud-user:231072:65536

cat /etc/subgid
lxd:165536:65536
root:165536:65536
cloud-user:231072:65536
```

2. 
```
cat ~/.config/lxc/default.conf 
lxc.include = /etc/lxc/default.conf
lxc.idmap = u 0 231072 65536
lxc.idmap = g 0 231072 65536
```

3.
```
lxc-usernsexec whoami
```
Expected result: root
Actual result:  `Failed to find subuid or subgid allocation`

**Comment by amarao** [comment]
Update: I bisect problem to upgrade between 3.0.1 and 3.0.3
Ubuntu bionic:
Version: `3.0.3-0ubuntu1~18.04.1` is affected
Version: `3.0.1-0ubuntu1~18.04.2` is not affected.

If someone find this bug before it was fixed, here ad-hoc solution:
```
apt-get install lxc-utils=3.0.1-0ubuntu1~18.04.2 liblxc-common=3.0.1-0ubuntu1~18.04.2 liblxc1=3.0.1-0ubuntu1~18.04.2
apt-mark hold lxc-utils liblxc-common  libxc1
```
(Do not forget to "unhold" packages when this bug is fixed).

**Comment by layer7gmbh** [comment]
Hi, 

same here, debian10

Version: 3.0.3

**Comment by anhlephuoc** [comment]
Hi,
Same problem in Debian 10.1 (LXC version 3.0.3). The failure is due to the failed lookup of default subuid/subgid.
Worked around without by using the -m option.
eg:  lxc-usernsexec -m 0:1000000:65536 -- touch /tmp/testfile.tmp
Hope that helps

**Comment by brauner** [comment]
This should be fixed in master.

**Comment by adavidson-tc** [comment]
Was this fixed in time for 3.0.4? Wondering if I should just put my hypervisor hosts onto Bullseye and take the later version as a fix.

**Comment by andydavidson** [comment]
Was this fixed in time for 3.0.4? Wondering if I should just put my hypervisor hosts onto Bullseye and take the later version as a fix?

**Comment by brauner** [comment]
@andydavidson I think this is fixed by:
```
commit c14ea11dccbfa80021a9b169b94bd86e8b359611
Author: Cameron Nemo <camerontnorman@gmail.com>
Date:   Wed Nov 28 19:42:29 2018 -0800

    lxc-usernsexec: fix default map functionality

    * Place NULL bytes at the end of strings so that
      lxc_safe_ulong() can parse them correctly

    * Only free the newly created id_map on error,
      to avoid passing garbage to lxc_map_ids()

    Signed-off-by: Cameron Nemo <camerontnorman@gmail.com>
```

Which seems to be in LXC 3.0.4.
